### PR TITLE
PLT-28: Payload logging in the APIs/gRPC interceptor

### PR DESCRIPTION
**Pull request by arkadysafety**
Sometimes it is useful to log out some of the payload details in a gRPC request. For example `id` or `userID`. This is useful for debugging and tracing of issues. However, We don't want to log all payload messages as 1) This would be a security leak and 2) Would clutter the logs with too verbose data.

The suggestion is to use a similar pattern to the Platform APIs validators: We create a field option in the protocol definitions that can state if a field should be logged. We can then create a generator that will generate an interface to log the payload with that field. This can be done in an interceptor similar to the Validate func. Eg

message User {
  string firstname = 1;
  string lastname = 2;
  string id = 3 [(logger.level) = "INFO"]

The idea here is that we can set a Log level that would be controllable via the logrus package based on the log level set on the application ENV.

### CODE REVIEW

**Comment by rmsafe** [comment]
Commit description is empty, (and as a follow on PR description is never autopopulated) :(.

**Comment by pawanrawal** [comment]
One more thing is that it looks like fields can be logged by giving them a special tag and using grpc tags. Was that looked into as it might be a simpler solution which we get out of the box? https://github.com/grpc-ecosystem/go-grpc-middleware/tree/master/tags#request-context-tags

**Comment by arkadysafety** [comment]
@pawanrawal,
1. If proto does not have log annotations output would still have file with Parse function that might be used to get logger parsed object with fields hidden depending on enabled logrus levels. This does not mean to be used though, function by default is called from grpc interceptor with may or may not be included into interceptor chain.
2. I have not done any benchmark tests, but performance would be as much slower as pretty much you create 1 extra variable each time interceptor runs + basic if checks. This would take milliseconds i would say.
3. Principle of this plugin exactly as it is for validators. Check examples for code. 

`request-context-tags` look to be different thing - an extra description per field that may be used while logging. What we try to build with this PR is hiding selected fields of the object while its being logged.

File: protobuf/protoc-gen-logger/plugin/plugin.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	p.s12proto = p.NewImport("github.com/SafetyCulture/s12-proto/protobuf/s12proto")
+
+	for _, msg := range file.Messages() {
+		p.generateParseFunction(file, msg)
+	}
+
+	p.generateLevelToLogrusFunction()
+}
+
+func (p *plugin) generateParseFunction(file *generator.FileDescriptor, message *generator.Descriptor) {
```
Comment (pawanrawal) [inline protobuf/protoc-gen-logger/plugin/plugin.go]: Can we write some tests for this to compare the output string?

File: protobuf/protoc-gen-logger/plugin/plugin.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	p.P(`case `, p.s12proto.Use(), `.Level_TRACE:`)
+	p.P(`return `, p.logrusPkg.Use(), `.TraceLevel`)
+	p.Out()
+	p.P(`}`)
+	p.P(`return 0`)
+	p.Out()
+	p.P(`}`)
+}
+
+func getLevelLteValue(field *descriptor.FieldDescriptorProto) *logger.Level {
```
Comment (pawanrawal) [inline protobuf/protoc-gen-logger/plugin/plugin.go]: Doesn't it just return the level value, why do you have `Lte` is that less than equal?

Comment (arkadysafety) [reply protobuf/protoc-gen-logger/plugin/plugin.go]: Good catch. Bad copy/paste.

**Pull request merged**